projectM_SDL_emscripten.cpp

```diff
@@ -254,35 +246,46 @@ int main(int argc, char* argv[])
 
     SDL_Init(SDL_INIT_VIDEO | SDL_INIT_AUDIO);
 
-    app.win = SDL_CreateWindow("projectM", 50, 50, width, height, 0);
-    app.rend = SDL_CreateRenderer(app.win, 0, SDL_RENDERER_ACCELERATED);
-    if (!app.rend)
+    SDL_CreateWindowAndRenderer(width, height, SDL_WINDOW_OPENGL, &app.win, &app.renderer);
+    if (!app.win || !app.renderer)
     {
-        fprintf(stderr, "Failed to create renderer: %s\n", SDL_GetError());
+        fprintf(stderr, "Failed to create SDL renderer: %s\n", SDL_GetError());
+        return 1;
+    }
+
+    SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION, "SDL renderer initialized.\n");
+
+    app.projectM = projectm_create();
+    if (!app.projectM) {
+        fprintf(stderr, "Failed to create projectM instance!\n");
         return 1;
     }
 
-    app.settings.mesh_x = 48;
-    app.settings.mesh_y = 48;
-    app.settings.fps = FPS;
-    app.settings.texture_size = 1024;
-    app.settings.window_width = width;
-    app.settings.window_height = height;
-    app.settings.soft_cut_duration = 3; // seconds
-    app.settings.preset_duration = 30; //seconds
-    app.settings.beat_sensitivity = 0.9;
-    app.settings.aspect_correction = true;
-    app.settings.easter_egg = 0; // ???
-    app.settings.shuffle_enabled = true;
-    app.settings.soft_cut_ratings_enabled = true; // ???
-    app.settings.preset_url = projectm_alloc_string(8);
-    strncpy(app.settings.preset_url, "/presets", 8);
-
-    // init projectM
-    app.pm = projectm_create_settings(&(app.settings), PROJECTM_FLAG_NONE);
-    projectm_select_random_preset(app.pm, true);
-    projectm_set_window_size(app.pm, width, height);
-    projectm_set_preset_switched_event_callback(app.pm, &presetSwitchedEvent, nullptr);
+    app.playlist = projectm_playlist_create(app.projectM);
+    if (!app.playlist)
+    {
+        fprintf(stderr, "Failed to create the playlist instance!\n");
+        return 1;
+    }
+
+    projectm_playlist_set_preset_switched_event_callback(app.playlist, &presetSwitchedEvent, nullptr);
+
+    projectm_set_mesh_size(app.projectM, 48, 32);
+    projectm_set_fps(app.projectM, FPS);
+    projectm_set_window_size(app.projectM, width, height);
+    projectm_set_soft_cut_duration(app.projectM, 3.0);
+    projectm_set_preset_duration(app.projectM, 5);
+    projectm_set_beat_sensitivity(app.projectM, 1.0);
+    projectm_set_aspect_correction(app.projectM, true);
+    projectm_set_easter_egg(app.projectM, 0);
+
+    projectm_set_preset_switch_failed_event_callback(app.projectM, &presetSwitchFailedEvent, nullptr);
```

Comment:
For now I'm redirecting the preset switch failed event to the app itself to prevent the playlist from spamming preset switches. The playlist manager will then only handle preset switch requests once.